### PR TITLE
WIP Feature/error handling and validation

### PR DESCRIPTION
**Pull request by genie9**
Validator's messages should be checked in validation functions: services/api_validator.py ???

There are a couple of comments in code with questions on how particular errors should be handled: services/api.py rows 898-901 and 1014-1016. DONE

Endpoint /v2/observation/ gives error: AttributeError: 'NoneType' object has no attribute 'serializer' because of unauthenticated request. How should this be handled? observations/api.py rows 23-24 doesn't seem to be the correct place, but from there empty serializer context is handed to observations/serializers.py row 127. TAKEN INTO CONSIDERATION

### CODE REVIEW

File: services/api.py [hunk 1]
```diff
@@ -32,6 +33,7 @@
 from rest_framework import renderers
 from django.template.loader import render_to_string
 from django.utils.module_loading import import_string
+from pydantic import ValidationError as pydantic_ValidationError
```
Comment (tituomin) [inline services/api.py]: I would just say

    import pydantic

here, and later use pydantic.ValidationError.

File: services/migrations/0068_servicenodeunitcount_city_as_department.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
```
Comment (tituomin) [inline services/migrations/0068_servicenodeunitcount_city_as_department.py]: This file should not be part of this PR.

File: services/api.py [hunk 3]
```diff
@@ -685,6 +687,15 @@ def get_queryset(self):
             queryset = queryset.prefetch_related('service_details__service')
 
         filters = self.request.query_params
+
+        # query parameter validation
+        f = {k: v[0].split(',') for k, v in filters.lists()}
+        try:
+            RequestFilters(**f)
+            UnitRequestFilters(**f)
+        except pydantic_ValidationError as e:
```
Comment (tituomin) [inline services/api.py]: See the import comment above. After changing that, I would just write

    except pydantic.ValidationError as e

File: services/api.py [hunk 4]
```diff
@@ -685,6 +687,15 @@ def get_queryset(self):
             queryset = queryset.prefetch_related('service_details__service')
 
         filters = self.request.query_params
+
+        # query parameter validation
+        f = {k: v[0].split(',') for k, v in filters.lists()}
```
Comment (tituomin) [inline services/api.py]: Two points.

First, we need to make sure that a comma is never part of a valid value. See SearchViewset for more info. It might be safer to only split the filters that explicitly support commas-as-separators.

Second, using the lists() method is a good idea, but it makes me realize that we should probably begin to support also  the &a=1&a=2&a=3 style for filters which support multiple values. That is, you could write something like

    {k: set([item for sublist in [x.split(',') for x in v] for item in sublist]) for k, v in d.items()}

But still the split(',') should only be done for supported fields.

This way, you could easily validate the fact that the API user has accidentally duplicated some parameter even though we don't support it (for that parameter).

Comment (tituomin) [reply services/api.py]: Also, you should only do the splitting once. Now it's done once for the validation and once below for the actual filter implementations.

File: services/api.py [hunk 5]
```diff
@@ -951,15 +960,17 @@ class SearchViewSet(munigeo_api.GeoModelAPIView, viewsets.ViewSetMixin, generics
     queryset = Unit.objects.all()
 
     def list(self, request, *args, **kwargs):
+        filters = self.request.query_params
+
         # If the incoming language is not specified, go with the default.
-        self.lang_code = request.query_params.get('language', LANGUAGES[0])
+        self.lang_code = filters.get('language', LANGUAGES[0])
         if self.lang_code not in LANGUAGES:
             raise ParseError("Invalid language supplied. Supported languages: %s" %
```
Comment (tituomin) [inline services/api.py]: I know this was ParseError in the code originally, but maybe it's not appropriate:

    Raised if the request contains malformed data when accessing request.data

Maybe a ValidationError could be better. However, could this be part of the pydantic validation too?

File: services/api.py [hunk 6]
```diff
@@ -973,8 +984,16 @@ def list(self, request, *args, **kwargs):
             else:
                 setattr(self, key, None)
 
-        input_val = request.query_params.get('input', '').strip()
-        q_val = request.query_params.get('q', '').strip()
+        # query parameter validation
+        f = {k: v[0].split(',') for k, v in filters.lists()}
```
Comment (tituomin) [inline services/api.py]: Especially here in the text search, where the input comes directly from the user, we cannot split all fields by commas -- the user can type a comma as part of `q` or `input` and it shouldn't be a problem...

File: services/api.py [hunk 7]
```diff
@@ -1071,12 +1091,16 @@ def list(self, request, *args, **kwargs):
...
         # Switch between paginated or standard style responses
         page = self.paginate_queryset(self.object_list)
-        serializer = self.get_serializer(page, many=True)
-        resp = self.get_paginated_response(serializer.data)
-
-        translation.activate(old_language)
-
-        return resp
+        serializer = self.get_serializer(page, many=True, data=request.data)
+        print(serializer)
```
Comment (tituomin) [inline services/api.py]: Remove.

Comment (genie9) [reply services/api.py]: This was left from testing and I think the check is not needed here. This part is 
restored to the original state.

File: services/api.py [hunk 8]
```diff
@@ -1009,48 +1054,42 @@ def list(self, request, *args, **kwargs):
...
 
-            if len(val) > 0:
-                deps_uuid = val.split(',')
+        if 'city_as_department' in filters:
+            deps_uuid = filters['city_as_department']
 
-                # forming municipality search query
-                deps = Department.objects.filter(uuid__in=deps_uuid).select_related('municipality')
-                munis = [d.municipality.name for d in deps]
+            # forming municipality search query
```
Comment (tituomin) [inline services/api.py]: These comments could be removed.

File: services/api_validator.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        if v not in levels:
+            raise ValueError('level must be one of: %s' % ', '.join(l for l in levels))
+
+    @validator('category', whole=True)
+    def check_category(cls, v):
+        for c in v:
+            if ':' not in c:
+                raise ValueError("'%s' not a valid category" % c)
+            else:
+                k = c.split(':')[0]
```
Comment (tituomin) [inline services/api_validator.py]: k, v = c.split(':')

File: services/api_validator.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                v = c.split(':')[1]
+                try:
+                    int(v)
+                except ValueError as e:
+                    raise ValueError('%s:%s, %s' % (k, v, str(e)))
+                if len(re.findall(r"\d", k)) > 0:
```
Comment (tituomin) [inline services/api_validator.py]: if k not in ['service', 'service_node']:

    ...

File: smbackend/exception_handler.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+def custom_exception_handler(exc, context):
+
+    # Call REST framework's default exception handler first,
+    # to get the standard error response.
+    response = exception_handler(exc, context)
+    json_resp = json.dumps(response.data) if response else exc
+
+    if isinstance(exc, ValidationError):
+        return HttpResponseBadRequest(content='ValidationError: {}'.format(json_resp),
```
Comment (tituomin) [inline smbackend/exception_handler.py]: Return a valid json object.

    error_json = {'ValidationError': response.data}
    error_txt = json.dumps(error_json)

File: services/api_validator.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+only_list = ['service_node.' + i for i in servicenode_fields] +\
+            ['service.' + i for i in service_fields] +\
+            ['unit.' + i for i in unit_fields]
+
+
+class RequestFilters(BaseModel):
+
+    page: int = None
+    page_size: int = None
+    include: List[str] = None
```
Comment (tituomin) [inline services/api_validator.py]: Remove

File: services/api_validator.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+    @validator('include', whole=True)
+    def include_search_params(cls, v):
+        if not any(p in include_list for p in v):
+            raise ValueError('include parameter must be one of: %s' % ', '.join(p for p in include_list))
+
+    @validator('only', whole=True)
+    def only_search_params(cls, v):
+        if not any(p in only_list for p in v):
+            raise ValueError('include parameter must be one of: %s' % ', '.join(p for p in unit_fields))
```
Comment (tituomin) [inline services/api_validator.py]: Use only_list

File: services/api.py [hunk 14]
```diff
@@ -747,19 +755,18 @@ def service_nodes_by_ancestors(service_node_ids):
...
 
+        service_nodes_excl = filters.get('exclude_service_nodes', None)
+
         service_node_ids = None
-        val = filters.get('exclude_service_nodes', None)
-        if val:
-            val = val.lower()
-            service_node_ids = val.split(',')
+        if service_nodes_excl:
+            service_node_ids = service_nodes_excl
```
[the remark this answers is not in the document]
Comment (genie9) [reply services/api.py]: this will change the logic of filter from:

`
service_nodes = filters.get(...) 
if service_nodes: 
    service_node_ids = service_nodes
elif level_specs: 
    if level_specs['type'] == 'include':
        service_node_ids = something
if service_node_ids: do something with service_node_ids
`

to:
`
service_nodes = filters.get(...)
if level_specs: 
    if level_specs['type'] == 'include':
        service_nodes = something
if service_nodes: do something with service_nodes
`

Is it ok that if query parameters have both: service_node and level with type include, then service_node ids will be overwritten?

File: services/api.py [hunk 15]
```diff
@@ -747,19 +755,18 @@ def service_nodes_by_ancestors(service_node_ids):
 
         service_node_ids = None
         if service_nodes:
-            service_nodes = service_nodes.lower()
-            service_node_ids = service_nodes.split(',')
+            service_node_ids = service_nodes
```
[the remark this answers is not in the document]
Comment (genie9) [reply services/api.py]: below question

File: services/api.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (tituomin) [reply services/api.py]: Yeah, good idea to add the `not service_nodes` condition.

File: smbackend/exception_handler.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+def custom_exception_handler(exc, context):
+
+    # Call REST framework's default exception handler first,
+    # to get the standard error response.
+    response = exception_handler(exc, context)
+    json_resp = json.dumps(response.data) if response else exc
+
+    if isinstance(exc, ValidationError):
+        return HttpResponseBadRequest(content=json.dumps('ValidationError: {}'.format(json_resp)),
```
Comment (tituomin) [inline smbackend/exception_handler.py]: This is not what we want. It technically does return valid JSON, but it's a single string! You're converting a python string to a json string. What we want is to convert a python dict to a json object.
